File: snippets/getting-started/token-refresh.py
"""Token refresh utilities for EVE SSO."""

import base64
import json
import time

import requests

SSO_TOKEN = "https://login.eveonline.com/v2/oauth/token"
USER_AGENT = "EVE Getting Started Guide/1.0 (your@email.com)"


def is_token_expired(access_token: str, buffer_seconds: int = 60) -> bool:
    """Check if a JWT access token is expired or will expire soon."""
    payload = access_token.split(".")[1]
    payload += "=" * (-len(payload) % 4)  # Fix base64 padding
    claims = json.loads(base64.urlsafe_b64decode(payload))
    return time.time() >= (claims["exp"] - buffer_seconds)


def refresh_access_token(
    refresh_token: str, client_id: str, client_secret: str
) -> dict:
    """Exchange a refresh token for a new access token."""
    response = requests.post(
        SSO_TOKEN,
        data={"grant_type": "refresh_token", "refresh_token": refresh_token},
        auth=(client_id, client_secret),
        headers={"User-Agent": USER_AGENT},
    )
    response.raise_for_status()
    return response.json()  # {"access_token": "...", "refresh_token": "...", ...}


def ensure_valid_token(
    access_token: str,
    refresh_token: str,
    client_id: str,
    client_secret: str,
) -> tuple[str, str]:
    """Return a valid access token, refreshing if needed."""
    if not is_token_expired(access_token):
        return access_token, refresh_token
    tokens = refresh_access_token(refresh_token, client_id, client_secret)
    return tokens["access_token"], tokens["refresh_token"]

File: snippets/getting-started/token-refresh.ts
// Token refresh utilities for EVE SSO.

const SSO_TOKEN = "https://login.eveonline.com/v2/oauth/token";
const USER_AGENT = "EVE Getting Started Guide/1.0 (your@email.com)";

/** Check if a JWT access token is expired or will expire soon. */
function isTokenExpired(accessToken: string, bufferSeconds = 60): boolean {
  const payload = JSON.parse(
    Buffer.from(accessToken.split(".")[1], "base64url").toString()
  );
  return Date.now() / 1000 >= payload.exp - bufferSeconds;
}

/** Exchange a refresh token for a new access token. */
async function refreshAccessToken(
  refreshToken: string,
  clientId: string,
  clientSecret: string
): Promise<{ access_token: string; refresh_token: string }> {
  const credentials = Buffer.from(`${clientId}:${clientSecret}`).toString(
    "base64"
  );
  const response = await fetch(SSO_TOKEN, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      Authorization: `Basic ${credentials}`,
      "User-Agent": USER_AGENT,
    },
    body: new URLSearchParams({
      grant_type: "refresh_token",
      refresh_token: refreshToken,
    }),
  });
  if (!response.ok) {
    throw new Error(`Token refresh failed: ${response.status}`);
  }
  return response.json();
}

/** Return a valid access token, refreshing if needed. */
async function ensureValidToken(
  accessToken: string,
  refreshToken: string,
  clientId: string,
  clientSecret: string
): Promise<[string, string]> {
  if (!isTokenExpired(accessToken)) {
    return [accessToken, refreshToken];
  }
  const tokens = await refreshAccessToken(refreshToken, clientId, clientSecret);
  return [tokens.access_token, tokens.refresh_token];
}
